### PR TITLE
Fix that RingBuffer does not get updated and add metrics about communication between endpoints

### LINKED ISSUE #5509: Add metrics about communication between endpoints
Hi,

I think it would be useful to have some information about how much data is transferred between Icinga 2 Endpoints, especially between masters and masters/satellites.

Could you add some metrics, so we can see a bit more what is going on between the different endpoins.
Current ideas for the API:
* rate of messages (increasing counter or per second) per connected endpoint
* rate of bytes (increasing counter or per secon) (please stick to bit or bytes and let users do the calculation for kilo/kibi, etc.) per connected endpoint
* last sent message per connected endpoit

Current ideas for the `icinga` Check:
* rate of messages or bytes (increasing counter or per second) total

rate per second would be easier to process for graphite while graphite and rrd can both handle increasing counters.

### LINKED ISSUE #5750: RingBuffer does not get updated if nothing is written
Currently entries only get removed when inserting new values to the buffer.

**Example of this problem:**
Lets say we're counting the messages that are sent from the master to another endpoint.
On connection the master might send a few messages but none after that.
If the master does not send messages for a few minutes, the buffer will not change for those minutes.

This would result in wrong performance data.

### PR DESCRIPTION
**Pull request by N-o-X**
fixes #5750 
fixes #5509

### CODE REVIEW

File: lib/base/ringbuffer.cpp [hunk 1]
```diff
@@ -58,14 +66,16 @@ void RingBuffer::InsertValue(RingBuffer::SizeType tv, int num)
 	m_Slots[offsetTarget] += num;
 }
 
-int RingBuffer::GetValues(RingBuffer::SizeType span) const
+int RingBuffer::UpdateAndGetValues(RingBuffer::SizeType tv, RingBuffer::SizeType span)
 {
+	InsertValue(tv, 0);
+
 	ObjectLock olock(this);
```
Comment (dnsmichi) [inline lib/base/ringbuffer.cpp]: Hm, shouldn't we lock the buffer before the insert happens?

File: lib/remote/apilistener.cpp [hunk 2]
```diff
@@ -1144,7 +1144,8 @@ void ApiListener::ReplayLog(const JsonRpcConnection::Ptr& client)
 				}
 
 				try  {
-					NetString::WriteStringToStream(client->GetStream(), pmessage->Get("message"));
+					size_t bytesSend = NetString::WriteStringToStream(client->GetStream(), pmessage->Get("message"));
+					endpoint->AddMessageSent(bytesSend);
```
Comment (dnsmichi) [inline lib/remote/apilistener.cpp]: I'm not happy with the naming schema. 

`message sent` does not directly tell you what it does. `bytesSend` should probably renamed to `sentBytes`.

This likely affects multiple code locations throughout.

File: lib/remote/apilistener.cpp [hunk 3]
```diff
@@ -1169,7 +1170,8 @@ void ApiListener::ReplayLog(const JsonRpcConnection::Ptr& client)
 					lmessage->Set("method", "log::SetLogPosition");
 					lmessage->Set("params", lparams);
 
-					JsonRpc::SendMessage(client->GetStream(), lmessage);
+					size_t bytesSend = JsonRpc::SendMessage(client->GetStream(), lmessage);
+					endpoint->AddMessageSent(bytesSend);
```
Comment (dnsmichi) [inline lib/remote/apilistener.cpp]: Similar, naming schema.

File: lib/remote/jsonrpcconnection.cpp [hunk 4]
```diff
@@ -111,7 +111,8 @@ void JsonRpcConnection::SendMessage(const Dictionary::Ptr& message)
 		ObjectLock olock(m_Stream);
 		if (m_Stream->IsEof())
 			return;
-		JsonRpc::SendMessage(m_Stream, message);
+		size_t bytesSend = JsonRpc::SendMessage(m_Stream, message);
+		m_Endpoint->AddMessageSent(bytesSend);
```
Comment (dnsmichi) [inline lib/remote/jsonrpcconnection.cpp]: Same.

File: lib/remote/apilistener.cpp (hunk 2 above)
Comment (N-o-X) [reply lib/remote/apilistener.cpp]: I already used bytesSent in other files, so I'll correct it to that.

**Comment by gunnarbeutner** [comment]
This introduces a new compiler warning:

```
[85/338] Building CXX object lib/remote/CMakeFiles/remote.dir/endpoint.cpp.o
../lib/remote/endpoint.cpp:38:27: warning: field 'm_BytesSent' will be initialized after field 'm_MessagesReceived' [-Wreorder]
    : m_MessagesSent(60), m_BytesSent(60), m_MessagesReceived(60), m_BytesReceived(60)
                          ^
1 warning generated.
```

**Pull request merged**